Sources/AardvarkReveal/ARKRevealArchiveBuilder.m

```diff
@@ -0,0 +1,285 @@
+                                       type:(char)type
+                                   fileSize:(int64_t)fileSize
+                                   linkName:(NSString *)linkName
+                                      error:(NSError **)error;
+{
+    // Create the header based on the Unix Standard TAR format.
+    // See <https://www.freebsd.org/cgi/man.cgi?query=tar&apropos=0&sektion=5&manpath=FreeBSD+7.0-RELEASE&arch=default&format=html>.
+    struct header_posix_ustar header = {
+        /* name */      "",
+        /* mode */      {'0', '0', '0', '7', '7', '7', ' ', '\0'},
+        /* uid */       {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* gid */       {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* size */      {'0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', ' '},
+        /* mtime */     {'0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', ' '},
+        /* checksum */  {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
+        /* typeflag */  type,
+        /* linkname */  "",
+        /* magic */     TMAGIC,
+        /* version */   TVERSION,
+        /* uname */     {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+        /* gname */     {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+        /* devmajor */  {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* devminor */  {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* prefix */    "",
+        /* pad */       ""
+    };
+
+    // Per the ustar spec, we should check the length of the file name and split it across the `name` and `prefix`
+    // fields if it's too long. In practice, we have a very shallow file hierarchy and short file names, so we should
+    // never overflow past the `name` field.
+
+    const char *_Nullable pathString = [path cStringUsingEncoding:NSUTF8StringEncoding];
+    memcpy(&header.name, pathString, [path lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
+
+    if (linkName != nil) {
+        const char *_Nullable linkString = [linkName cStringUsingEncoding:NSUTF8StringEncoding];
+        memcpy(&header.linkname, linkString, [linkName lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
+    }
+
+    int failed = 0;
+
+    failed |= write_ascii_octal(fileSize, (char *)(&header.size), sizeof(header.size) - 1);
+
+    failed |= write_ascii_octal(_modificationDate, (char *)(&header.mtime), sizeof(header.mtime) - 1);
+
+    int64_t checksum = 0;
+    for (int i = 0 ; i < TAR_BLOCK_LENGTH ; i++) {
+        checksum += 255 & (int64_t)(((char *)&header)[i]);
+    }
+    failed |= write_ascii_octal(checksum, (char *)&header.checksum, 6);
+
+    if (failed) {
+        *error = [NSError errorWithDomain:ARKRevealArchiveBuilderErrorDomain
+                                     code:ARKRevealArchiveBuilderErrorCodeHeaderWriteFailure
+                                 userInfo:nil];
+    }
+
+    return header;
+}
```

Comment:
I'd feel more confident reviewing this code if this code had unit tests.

Reply:
Added some tests in 462da40